### ISSUE TITLE
Excessive memory consumption on config-reload

### ISSUE THREAD

**Issue by dbaumgarten** [nginxinc/kubernetes-ingress#3907]
**Describe the bug**
When a deployment in Kubernetes is updated, this triggers repeated config-reloads in nginx (whenever a pod becoms (un)ready).
During these repeated config-reloads the memory consumption of the nginx-ingress-controller raises drastically.

For example: From 480MB to over 5 GB.

I found https://github.com/nginxinc/kubernetes-ingress/issues/1116 which might be related.
My monitoring shows that during the reloads the number of active worker-processes raises from the default 8 to over 60.

I understand that the increase in memory consumption is related to the number of worker-processes.

However the amount of memory-usage seems quite drastic to me. That is a ten-fold increase. For something as simple as performing a rolling-update to 3 pods.

The traffic during the restart was relatively light (6am...) and our workload has practically no long-running HTTP-Connections (no websockets, no streaming etc.).

It this behavior normal?
Can I do anything to mitigate this? 

**To Reproduce**
1. Have a demo-deployment with a few replicas
2. Have some light traffic to these demo-pods over the nginx-ingress
3. Trigger a rolling-update of the demo-deployment
4. Observe the memory-usage of nginx-ingress

**Expected behavior**
Resource-usage of nginx should not dramatically increase during normal cluster-operations.

**Your environment**
* Version of the Ingress Controller - release version or a specific commit: 3.1.1
* Version of Kubernetes: 1.23.17
* Kubernetes platform (e.g. Mini-kube or GCP): EKS
* Using NGINX 

I have attached a screensot of the memory-consumption and the config that is used for the ingress-controller.
I can provide more metrics/logs etc. if needed.

![screenshot](https://github.com/nginxinc/kubernetes-ingress/assets/8211271/20adb0ad-cd7a-4033-ae6e-f7c306bedd33)
[config.txt](https://github.com/nginxinc/kubernetes-ingress/files/11497243/config.txt)

**Comment by shaun-nx** [comment]
Hi @dbaumgarten thank you for reporting this! I'm looking into this right now. I'll get back to you as soon as we discover the root cause.

**Comment by shaun-nx** [comment]
Hi @dbaumgarten can you provide the deployment yaml of your demo-deployment? Thanks.

**Comment by dbaumgarten** [comment]
The workload we are seing this behavior with is a little to complicated and has several dependencies.
I will try to reproduce this with a simpler workload and provide that yaml here

**Comment by jjngx** [comment]
@dbaumgarten would you be able to provide info about how much memory is used by NGINX proces(es)?

**Comment by dbaumgarten** [comment]
Providing the total memory-usage of the pod/container is easy (see inital post).
But I have no idea how to get this information on process-level...

**Comment by dbaumgarten** [comment]
I have investigated this a little more and got to the following insights:
1. This only happens on clusters where our "main"-workload is running (so not a general problem with the ingress-controller)
2. It does not matter which deployment is updated, as long as the update-workload and our "main"-workload are served by the same ingress-controller
3. The memory-increase happens even if there is low (practically no) traffic on the controller. However, if there is traffic on the controller (even if just light traffic) the memory-increase is far more drastical


I reproduced this on our dev-cluster:
In this cluster are two nginx-ingress-controllers. Both forward to the same pods, but handle incoming traffic from different networks.

- The "internal"-ingress-controller on this clusters serves our continously running smoke-test. As this is the dev-cluster no other traffic is served, at all.
As you see in this graph, the number of active connections is pretty low:
![intraned_connections](https://github.com/nginxinc/kubernetes-ingress/assets/8211271/44cef0d6-45d6-4441-b79b-f8e1b7372c83)

However, despite the low traffic-volume, memory-consumption spikes massively during a rolling-update of a deployment:
![intranet_memory](https://github.com/nginxinc/kubernetes-ingress/assets/8211271/a83bf554-3f83-4ea1-9b9b-c8b72045c4e8)

During that time the amount of worker-processes also spikes up:
![workers_intranet](https://github.com/nginxinc/kubernetes-ingress/assets/8211271/8ed20275-f6d7-43c1-941b-18796dc7edc7)


- On the "external"-ingress-controller, which handles practically no traffic at all on this cluster (1-2 active connctions, which are probably just healthchecks), the memory-usage during the rolling-update of a deployment "only" goes up from 500mb to 1.2GB


This makes me think it might have something to do with the nature of our traffic. (Which in this case are only automated tests with python/selenium, which should just take <1s per HTTP-Call)

But the fact that the memory-consumption of the "external"-ingress-controller more than doubles, even though it is serving no traffic at all, really confuses me.

Any ideas?

**Comment by dbaumgarten** [comment]
I just tested it with v3.0.0.
The behavior is exactly the same: 318mb before and 3.5GB during the reload.

**Comment by jjngx** [comment]
> I just tested it with v3.0.0.
> The behavior is exactly the same: 318mb before and 3.5GB during the reload.

Thank you @dbaumgarten for running the test and providing feedback. cc/ @shaun-nx

**Comment by dbaumgarten** [comment]
You're welcome.
If there is anything else I can test, please let me know!

**Comment by shaun-nx** [comment]
Hi @dbaumgarten thank you so much for doing this testing. Right now with the cluster setups we are using we have not been able to reproduce this memory spike. We'd like to try to align our environment with yours to make testing more accurate on our end.

Can you provide details of your dev environment you used for testing? Things like the number of nodes and the specifications of those nodes (CPU, Memory)

You mentioned that you're doing a rolling-updated to your applications. What kind of image are you using on these applications? Is it something that can be shared with us? If not that is perfectly fine. Even details such as the size of that image will help.

The last thing I'd like to ask is can you run this same test against v2.4.2? At the moment we suspect that this issue is caused by our implementation of EndpointSlices which was added in v3.0.0.

Thank you again.

**Comment by shaun-nx** [comment]
@dbaumgarten can you also provide details on the tools you are using to monitor the memory increase and the amount of worker-processes? It would help to make sure we are measuring the statistics the same way. Thanks!

**Comment by dbaumgarten** [comment]
Hi,
I have investigated this a little mor on my side and I think I found the issue.
It was "broken OCSP-Stapling".

The ingress-controller was configured to perform ocsp-stapling. But due to a recent networking-change direct connections to the internet are no longer possible. This resulted in the ocsp-calls timing out.

So here is what I think happened:
1. A kubernetes deployment-rollout is triggering repeated reloads in a short time
2. New worker-processes are spawned by nginx
3. Each worker-process is trying to perform ocsp-calls
4. Nginx tries to shut down old workers. However as they are still waiting for the ocsp-response, they keep running.
5. Old worker-processes (which are still active because they are waiting for the ocsp-response) start piling up
6. Now there are so many running worker-processes that it really raises the memory-requirement
7. Now the ocsp-calls start to time-out and workers actually shut down -> memory usage normalizes

Tl;dr -> Waiting for OCSP-Responses keeps old workers busy. Many new workers are started because of repeated config-reloads -> many simultaneous workers eat loads of memory.

![img](https://github.com/nginxinc/kubernetes-ingress/assets/8211271/9be7aa2e-f356-4cfb-9fae-3803c118c7d5)
At ~6:45 I triggered a redeployment with brokern ocsp-stapling enabled.
At 7:04 I disabled ocsp-stapling.
At 7:07 and 7:15 I triggered redployments. (Now with ocsp-stapling disabled)

So in the end it seems it was not a bug in the ingress-controller but just poor configuration on my side.
I am sorry for having annoyed you with my stupidness...

**Comment by shaun-nx** [comment]
Hi @dbaumgarten I'm glad you were able to discover the cause! And please don't feel sorry at all! 😄

**Comment by vepatel** [comment]
Hi @dbaumgarten, can you still please let us know what tooling and dashboard you've used to measure memory consumption!

**Comment by dbaumgarten** [comment]
Sure. I am using Prometheus and it's node-exporter to collect data about the memory-consumption.
Number of active workers are gathered via Prometheus from the metrics exported by the ingress-controller.
The data from prometheus is visualized in grafana.
I am using the kube-pronetheus stack which conveniently bundles Prometheus, grafana and lots of monitoring-rules.
